src/Parser.js

```diff
@@ -461,7 +463,7 @@ export default class Parser {
             const result = env.handler(context, args, optArgs);
             this.expect("\\end", false);
             const endNameToken = this.nextToken;
-            const end = this.parseFunction();
+            const end = assertNodeType(this.parseFunction(), "environment");
```

Comment:
Ah, it was not until merging with my PR that I realized. I'd done this `assertNodeType` after the `if (!end)` below so that we could give a more logical error message. What do you think? I can leave that in in my current PR. (We can discuss it on my PR, I suppose.)